overlord/configstate/configcore/kernel.go

```diff
@@ -0,0 +1,213 @@
+	coreOptionKernelDangerousCmdlineAppend = "core.system.kernel.dangerous-cmdline-append"
+)
+
+func init() {
+	supportedConfigurations[coreOptionKernelCmdlineAppend] = true
+	supportedConfigurations[coreOptionKernelDangerousCmdlineAppend] = true
+}
+
+func changedKernelConfigs(c RunTransaction) []string {
+	changed := []string{}
+	for _, name := range c.Changes() {
+		// Note that we cannot just check the prefix as we have
+		// system.kernel.* options also defined in sysctl.go.
+		if name == coreOptionKernelCmdlineAppend || name == coreOptionKernelDangerousCmdlineAppend {
+			nameWithoutSnap := strings.SplitN(name, ".", 2)[1]
+			changed = append(changed, nameWithoutSnap)
+		}
+	}
+	return changed
+}
+
+func validateCmdlineParamsAreAllowed(st *state.State, devCtx snapstate.DeviceContext, params []string) error {
+	// gd, err := devicestate.CurrentGadgetInfo(st, devCtx)
+	// if err != nil {
+	// 	return err
+	// }
+	// logger.Debugf("gadget data read from %s", gd.RootDir)
+	// TODO use gadgetdata to check against allowed values
+
+	return nil
+}
+
+func validateCmdlineAppend(c RunTransaction) error {
+	changed := changedKernelConfigs(c)
+	if len(changed) == 0 {
+		return nil
+	}
+
+	st := c.State()
+	st.Lock()
+	defer st.Unlock()
+	devCtx, err := devicestate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	for _, opt := range changed {
+		cmdAppend, err := coreCfg(c, opt)
+		if err != nil {
+			return err
+		}
+
+		logger.Debugf("validating %s=%q", opt, cmdAppend)
+		params, err := osutil.KernelCommandLineSplit(cmdAppend)
+		if err != nil {
+			return err
+		}
+		if opt == optionKernelCmdlineAppend {
+			// check against allowed values from gadget
+			if err := validateCmdlineParamsAreAllowed(c.State(), devCtx, params); err != nil {
+				return err
+			}
+		} else { // OptionKernelDangerousCmdlineAppend
+			if devCtx.Model().Grade() != asserts.ModelDangerous {
+				// TODO we should return an error if this is an API call
+				// and do nothing if setting defaults (so gadget can be
+				// reused with different models).
+				logger.Noticef("WARNING: %s ignored as this is not a dangerous model", opt)
+			}
+		}
+	}
+
+	return nil
+}
+
+func createApplyCmdlineChange(c RunTransaction, kernelOpts []string) (*state.Change, error) {
+	st := c.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// error out if some other change is touching the kernel command line
+	if err := snapstate.CheckUpdateKernelCommandLineConflict(st, ""); err != nil {
+		return nil, err
+	}
+
+	// precalculate task arguments, so we do not need to destroy change/task
+	// if there is an error
+	args := []struct {
+		name    string
+		cmdline string
+	}{}
+	for _, opt := range kernelOpts {
+		cmdline, err := coreCfg(c, opt)
+		if err != nil {
+			return nil, err
+		}
+		// take just *cmdline-append (it should not fail)
+		name := strings.Split(opt, ".")[2]
+		args = append(args, struct {
+			name    string
+			cmdline string
+		}{
+			name: name, cmdline: cmdline})
```

Comment:
(super nitpick) I would move the closing into the next line to make things more symetric when glancing over the page.
```go
@@ -141,7 +141,8 @@ func createApplyCmdlineChange(c RunTransaction, kernelOpts []string) (*state.Cha
                        name    string
                        cmdline string
                }{
-                       name: name, cmdline: cmdline})
+                       name: name, cmdline: cmdline,
+               })
        }
 
        // We need to create a new change that will change the kernel
```

Reply:
Done